akka-docs/rst/scala/distributed-data.rst

```diff
@@ -0,0 +1,387 @@
+from a database record that is used for optimistic concurrency control.
+
+For first-write-wins semantics you can use the ``LWWRegister#reverseClock`` instead of the
+``LWWRegister#defaultClock``.
+
+Custom Data Type
+----------------
+
+You can rather easily implement your own data types. The only requirement is that it implements
+the ``merge`` function of the ``ReplicatedData`` trait.
+
+A nice property of stateful CRDTs is that they typically compose nicely, i.e. you can combine several
+smaller data types to build richer data structures. For example, the ``PNCounter`` is composed of
+two internal ``GCounter`` instances to keep track of increments and decrements separately.
+
+FIXME code example of "simplified" PNCounter
+
+FIXME perhaps show the shopping cart here?
+
+Data types should be immutable, i.e. "modifying" methods should return a new instance.
+
+The data types must be serializable with an :ref:`Akka Serializer <serialization-scala>`.
+It is highly recommended that you implement  efficient serialization with Protobuf or similar
+for your custom data types. The built in data types are marked with ``ReplicatedDataSerialization``
+and serialized with ``akka.cluster.ddata.protobuf.ReplicatedDataSerializer``.
+
+Serialization of the data types are used in remote messages and also for creating message
+digests (SHA-1) to detect changes. Therefore it is important that the serialization is efficient
+and produce the same bytes for the same content. For example sets and maps should be sorted
+deterministically in the serialization.
+
+FIXME code example of serializer
+
+CRDT Garbage
+------------
+
+One thing that can be problematic with CRDTs is that some data types accumulate history (garbage).
+For example a ``GCounter`` keeps track of one counter per node. If a ``GCounter`` has been updated
+from one node it will associate the identifier of that node forever. That can become a problem
+for long running systems with many cluster nodes being added and removed. To solve this problem
+the ``Replicator`` performs pruning of data associated with nodes that have been removed from the
+cluster. Data types that need pruning have to implement the ``RemovedNodePruning`` trait. 
+
+Samples
+=======
+
+FIXME activator template with the samples
+
+* Replicated Cache
+* Replicated Metrics
+* Replicated Service Registry
+* VotingService
+* ShoppingCart 
+
+Limitations
+===========
+
+There are some limitations that you should be aware of.
+
+CRDTs cannot be used for all types of problems, and eventual consistency does not fit
+all domains. Sometimes you need strong consistency.
+
+It is not intended for *Big Data*. The number of top level entries should not exceed 100000.
+When a new node is added to the cluster all these entries are transferred (gossiped) to the
+new node. The entries are split up in chunks and all existing nodes collaborate in the gossip,
+but it will take a while (tens of seconds) to transfer all entries and this means that you 
```

Comment:
I'm sure there are several ways that we can improve the sync of new nodes, but it is pretty good as is, and we can do that as a separate improvement later, when needed.